### ISSUE TITLE
Unexpected error occurred with message: Could not retreive AWS credentials, credential value is nil or empty.

### ISSUE THREAD

**Issue by ethan021021** [aws-amplify/amplify-swift#2527]
### Describe the bug

Hello,

Currently in our app we are receiving errors logged from users who seem to be running into this very random auth issue:

```
APIError: Failed to retrieve authorization token.
Caused by:
AuthError: Unexpected error occurred with message: Could not retreive AWS credentials, credential value is nil or empty.
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: /Users/username/Library/Developer/Xcode/DerivedData/Serenity-ammhovqgzfuogudxbhivmwhannha/SourcePackages/checkouts/amplify-ios/Amplify/Categories/Auth/Error/AuthError.swift
function: recoverySuggestion
line: 80
```

This is effecting 1/3rd of our user base and these users are unable to make calls using the API category while logged in through the Auth category. Our Auth is using Cognito user pools and hasn't given us much of a headache until this issue started to pop up. From the error's being returned in our error reporting service (Sentry) an API call isn't even made I assume `toAmplifyAWSCredentials()` throws and stops execution when trying to make the network call. 

I've tried countless times to reproduce this issue myself but have been unable to do so. After looking through the Amplify swift package the issue might have to do with the `toAmplifyAWSCredentials()` function as `accessKey` and `secretKey` seem to be `nil` in this specific case. What I couldn't figure out is if the code path to `toAmplifyAWSCredentials()` lies within `AuthorizationProviderAdapter+SignedInSession` or `AuthorizationProviderAdapter+SignedOutSession`. What is the difference between these two? Any additional information here would be great so I can figure out how the `accessKey` and `secretKey` aren't getting set. If there's any additional information I can provide to help track down this issue please do let me know!

As asked for in the Discord:

Configuration File (NOTE: Sign in with Apple is currently included in our awsconfiguration but is not currently in use):
```
{
    "UserAgent": "aws-amplify/cli",
    "Version": "0.1.0",
    "IdentityManager": {
        "Default": {}
    },
    "CredentialsProvider": {
        "CognitoIdentity": {
            "Default": {
                "PoolId": "us-west-1:{REDACTED}",
                "Region": "us-west-1"
            }
        }
    },
    "CognitoUserPool": {
        "Default": {
            "PoolId": "us-west-1_{REDACTED}",
            "AppClientId": "{REDACTED}",
            "Region": "us-west-1"
        }
    },
    "Auth": {
        "Default": {
            "OAuth": {
                "WebDomain": "{REDACTED}",
                "AppClientId": "{REDACTED}",
                "SignInRedirectURI": "{REDACTED}",
                "SignOutRedirectURI": "{REDACTED}",
                "Scopes": [
                    "phone",
                    "email",
                    "openid",
                    "profile",
                    "aws.cognito.signin.user.admin"
                ]
            },
            "authenticationFlowType": "USER_SRP_AUTH",
            "socialProviders": [
                "APPLE"
            ],
            "usernameAttributes": [
                "EMAIL"
            ],
            "signupAttributes": [
                "EMAIL"
            ],
            "passwordProtectionSettings": {
                "passwordPolicyMinLength": 8,
                "passwordPolicyCharacters": []
            },
            "mfaConfiguration": "OFF",
            "mfaTypes": [
                "SMS"
            ],
            "verificationMechanisms": [
                "EMAIL"
            ]
        }
    },
    "AppSync": {
        "Default": {
            "ApiUrl": "{REDACTED}",
            "Region": "us-west-1",
            "AuthMode": "AMAZON_COGNITO_USER_POOLS",
            "ClientDatabasePrefix": "serenity_AMAZON_COGNITO_USER_POOLS"
        },
        "serenity_AWS_IAM": {
            "ApiUrl": "{REDACTED}",
            "Region": "us-west-1",
            "AuthMode": "AWS_IAM",
            "ClientDatabasePrefix": "serenity_AWS_IAM"
        }
    },
    "S3TransferUtility": {
        "Default": {
            "Bucket": "{REDACTED}",
            "Region": "us-west-1"
        }
    },
    "PinpointAnalytics": {
        "Default": {
            "AppId": "{REDACTED}",
            "Region": "us-west-2"
        }
    },
    "PinpointTargeting": {
        "Default": {
            "Region": "us-west-2"
        }
    }
}

```

### Steps To Reproduce

```swift
N/A
```

### Expected behavior

User is able to make API calls through the API Category while authenticated with the Auth Category

### Amplify Framework Version

1.28.2

### Amplify Categories

API, Auth

### Dependency manager

Swift PM

### Swift version

5.6

### CLI version

10.4.0

### Xcode version

14.0

### Relevant log output

```shell
APIError: Failed to retrieve authorization token.
Caused by:
AuthError: Unexpected error occurred with message: Could not retreive AWS credentials, credential value is nil or empty.
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: /Users/ethanthomas/Library/Developer/Xcode/DerivedData/Serenity-ammhovqgzfuogudxbhivmwhannha/SourcePackages/checkouts/amplify-ios/Amplify/Categories/Auth/Error/AuthError.swift
function: recoverySuggestion
line: 80
```

### Is this a regression?

No

### Device

iPhone SE (2nd generation)

### iOS Version

16.0.3

**Comment by ethan021021** [comment]
Also after some searching it looks like this is the exact problem I'm running into: https://github.com/aws-amplify/amplify-swift/issues/1501

**Comment by michaeljajou** [comment]
I am also experiencing this - any update on this issue?

**Comment by michaeljajou** [comment]
This issue is also mentioned here: https://github.com/aws-amplify/amplify-flutter/issues/2312

This seems like an urgent issue and should receive attention immediately.

**Comment by atierian** [comment]
@ethan021021 did this just start happening with 1.28.2?

@michaeljajou can you please confirm which version of Amplify you're experiencing this with? And are you able to reliably reproduce the issue? If so, please post the steps you're taking to do so. Either way, please post your amplifyconfiguration.json (with any confidential information redacted). That will help us in attempts to reproduce.

We're going to take a close look at this tomorrow and I, or someone else, will get back to you here with an update.

**Comment by michaeljajou** [comment]
@atierian 

Amplify (1.28.1):
    - Amplify/Default (= 1.28.1)
  - Amplify/Default (1.28.1)
  - AmplifyPlugins/AWSCognitoAuthPlugin (1.28.1):
    - AWSAuthCore (~> 2.28.0)
    - AWSCognitoIdentityProvider (~> 2.28.0)
    - AWSCognitoIdentityProviderASF (~> 2.28.0)
    - AWSCore (~> 2.28.0)
    - AWSMobileClient (~> 2.28.0)
    - AWSPluginsCore (= 1.28.1)
  - AWSAuthCore (2.28.0):
    - AWSCore (= 2.28.0)
  - AWSCognitoIdentityProvider (2.28.0):
    - AWSCognitoIdentityProviderASF (= 2.28.0)
    - AWSCore (= 2.28.0)
  - AWSCognitoIdentityProviderASF (2.28.0):
    - AWSCore (= 2.28.0)
  - AWSCore (2.28.0)
  - AWSMobileClient (2.28.0):
    - AWSAuthCore (= 2.28.0)
    - AWSCognitoIdentityProvider (= 2.28.0)
    - AWSCognitoIdentityProviderASF (= 2.28.0)
    - AWSCore (= 2.28.0)
  - AWSPluginsCore (1.28.1):
    - Amplify (= 1.28.1)
    - AWSCore (~> 2.28.0)

**Comment by michaeljajou** [comment]
@atierian i haven't been able to reproduce the issue myself, I've only heard it back from a user yesterday. He purchased a new phone, restored from iCloud, and since has been unable to login/create an account with the error showing "an unexpected error has occurred" 

So definitely something going on in the auth category.

**Comment by atierian** [comment]
Ok, thanks. That extra context is helpful. We're looking into this - if you discover anything else or have any other thoughts, please keep us posted here.

**Comment by michaeljajou** [comment]
Awesome thanks - will do. Another piece that may be helpful was that the user was able to successfully reset their password

**Comment by atierian** [comment]
Thanks for your patience. We were able to reproduce this issue by fetching the auth session while invalidating the credential cache in the SDK, which is something that can occur in the wild. We have a fix for it here https://github.com/aws-amplify/aws-sdk-ios/pull/4379 and will be cutting a release soon. Once the SDK is released, you'll be able to benefit from the change without updating Amplify. Let us know if you have any questions, thanks!

**Comment by michaeljajou** [comment]
Awesome news, thank you! I am still on Amplify 2.28.0 - so I will not need to release a new app version with a later version of Amplify you're saying?

**Comment by atierian** [comment]
You will need to release a new app version. Amplify uses _up to next minor_ versioning for the underlying AWS SDK for iOS (what was fixed). You will likely need to resolve package versions for SPM or `pod install` for Cocoapods, and confirm that it is now using the correct version of the AWS SDK for iOS. We will update here once the release is complete.

**Comment by ethan021021** [comment]
Hi thanks for the quick fix! Will the fix also work for Amplify version `1.28.2`? Thank you

**Comment by atierian** [comment]
Yes, it will. Just make sure that the AWS SDK for iOS dependency is the new version (2.28.2) once it's released.

**Comment by michaeljajou** [comment]
Any ETA when this will be released?

**Comment by atierian** [comment]
[AWS SDK for iOS v2.28.2](https://github.com/aws-amplify/aws-sdk-ios/releases/tag/2.28.2) was just released and includes this fix. You can now take the steps I mentioned a few comments up - just please double check that it's actually using v2.28.2 before pushing a new release of your app. Thanks again for bringing this issue to our attention.

I'm closing this issue now. If you have any other questions, concerns, or run into any problems related to this, please comment on this issue.

**Comment by michaeljajou** [comment]
@atierian I've heard back from a user that this is still happening - but I have a feeling they haven't updated, waiting for a confirmation from them. 

Although - I am now on Amplify v2.28.3 - I want to confirm that this fix still applies in that version? 

`PODS:
  - Amplify (1.28.2):
    - Amplify/Default (= 1.28.2)
  - Amplify/Default (1.28.2)
  - AmplifyPlugins/AWSCognitoAuthPlugin (1.28.2):
    - AWSAuthCore (~> 2.28.0)
    - AWSCognitoIdentityProvider (~> 2.28.0)
    - AWSCognitoIdentityProviderASF (~> 2.28.0)
    - AWSCore (~> 2.28.0)
    - AWSMobileClient (~> 2.28.0)
    - AWSPluginsCore (= 1.28.2)
  - AWSAuthCore (2.28.3):
    - AWSCore (= 2.28.3)
  - AWSCognitoIdentityProvider (2.28.3):
    - AWSCognitoIdentityProviderASF (= 2.28.3)
    - AWSCore (= 2.28.3)
  - AWSCognitoIdentityProviderASF (2.28.3):
    - AWSCore (= 2.28.3)
  - AWSCore (2.28.3)
  - AWSMobileClient (2.28.3):
    - AWSAuthCore (= 2.28.3)
    - AWSCognitoIdentityProvider (= 2.28.3)
    - AWSCognitoIdentityProviderASF (= 2.28.3)
    - AWSCore (= 2.28.3)
  - AWSPluginsCore (1.28.2):
    - Amplify (= 1.28.2)
    - AWSCore (~> 2.28.0)
  - mamba (2.2.1)
  - SwiftCheck (0.12.0)
  - SwiftLint (0.49.1)
  - SwiftUIPager (2.5.0)`

**Comment by SESA762550** [comment]
@atierian I am facing the same issue even in the  version AWSMobileClient 2.34.2, I am using Amplify 1.30.8
Is there anything else that needs to be done to resolve this issue ?

**Comment by atierian** [comment]
Hi @SESA762550, thanks for reporting this. Please open a new issue including the error message, your amplify configuration, and any logs. Thanks!
